### PR TITLE
Implement the create an element for token algorithm

### LINKED ISSUE #19392: Implement create-an-element-for-a-token algorithm
`create_html_element` implements https://dom.spec.whatwg.org/#concept-create-element and is called from the parser (via Element::create in `Sink::create_element`), but the HTML specification defines https://html.spec.whatwg.org/multipage/parsing.html#create-an-element-for-the-token which is used by the HTML parsing algorithms. Steps 6 and 9 are particularly relevant for dealing with https://github.com/servo/servo/pull/19307#issuecomment-346173603.

### LINKED ISSUE #19393: Implement throw-on-dynamic-markup-insertion value in document
Spec: https://html.spec.whatwg.org/multipage/dynamic-markup-insertion.html#throw-on-dynamic-markup-insertion-counter

There are TODO lines in document.rs about dealing with this properly. This is required for resolving https://github.com/servo/servo/pull/19307#issuecomment-346173603 properly, since #19392 requires it.

### PR DESCRIPTION
**Pull request by cbrewster**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #19392 and fix #19393 (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19397)

### CODE REVIEW

**Comment by cbrewster** [comment]
r? @jdm 
Was there a specific test case that I could add here?
Also do we have a way of determining if the JavaScript execution context stack is empty?

**Comment by jdm** [comment]
The test that led me to discover this was to take parser-fallsback-to-unknown-element.html and add a `document.write()` call in an `onerror` handler. Without #19393 the parser is in an unexpected state and we get a borrow panic.

File: components/script/dom/servoparser/mod.rs [hunk 1]
```diff
@@ -946,3 +956,61 @@ impl TreeSink for Sink {
...
+    // Step 5.
+    // TODO: Check if the parser was created for the HTML fragment parsing algorithm
+    let will_execute_scrtipt = definition.is_some();
+
+    // Step 6.
+    if will_execute_scrtipt {
+        // Step 6.1.
+        // TODO: handle throw-on-dynamic-markup-insertion counter.
+        // Step 6.2
+        // TODO: If the JavaScript execution context stack is empty, then perform a microtask checkpoint.
```
Comment (jdm) [inline components/script/dom/servoparser/mod.rs]: We should add an API to https://github.com/servo/servo/blob/master/components/script/dom/bindings/settings_stack.rs to check this.

File: components/script/dom/servoparser/mod.rs [hunk 2]
```diff
@@ -946,3 +956,61 @@ impl TreeSink for Sink {
...
+    let is = attrs.iter()
+        .find(|attr| attr.name.local.eq_str_ignore_ascii_case("is"))
+        .map(|attr| LocalName::from(&*attr.value));
+
+    // Step 4.
+    let definition = document.lookup_custom_element_definition(&name.ns, &name.local, is.as_ref());
+
+    // Step 5.
+    // TODO: Check if the parser was created for the HTML fragment parsing algorithm
+    let will_execute_scrtipt = definition.is_some();
```
Comment (jdm) [inline components/script/dom/servoparser/mod.rs]: `script`

File: components/script/dom/servoparser/mod.rs [hunk 3]
```diff
@@ -946,3 +956,61 @@ impl TreeSink for Sink {
...
+    // Step 3.
+    let is = attrs.iter()
+        .find(|attr| attr.name.local.eq_str_ignore_ascii_case("is"))
+        .map(|attr| LocalName::from(&*attr.value));
+
+    // Step 4.
+    let definition = document.lookup_custom_element_definition(&name.ns, &name.local, is.as_ref());
+
+    // Step 5.
+    // TODO: Check if the parser was created for the HTML fragment parsing algorithm
```
Comment (jdm) [inline components/script/dom/servoparser/mod.rs]: We should add a field to ServoParser with this information.

File: components/script/dom/servoparser/mod.rs [hunk 4]
```diff
@@ -946,3 +956,61 @@ impl TreeSink for Sink {
...
+    if will_execute_scrtipt {
+        // Steps 9.1 - 9.2.
+        ScriptThread::pop_current_element_queue()
+        // Step 9.3.
+        // TODO: handle throw-on-dynamic-markup-insertion counter.
+    }
+
+    // TODO: Step 10.
+    // TODO: Step 11.
+    // TODO: Step 12.
```
Comment (jdm) [inline components/script/dom/servoparser/mod.rs]: Step 12 is handled from the parser callback `associate_with_form`.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by jdm** [comment]
Bleah:
```
./components/script/dom/servoparser/mod.rs:110: derivable traits list is not in alphabetical order
	expected: Clone, Copy, JSTraceable, MallocSizeOf, PartialEq
	found: Clone, Copy, PartialEq, JSTraceable, MallocSizeOf
```

**Comment by CYBAI** [comment]
Failed with #19495

**Comment by jdm** [comment]
I appreciate the support, but I was watching the build and it didn't need a retry.

**Pull request merged**